src/Games/NexusMods.Games.FOMOD/CoreDelegates/UiDelegate.cs

```diff
@@ -69,74 +79,130 @@ public void EndDialog()
+
+                    var groupings = selectedOptions.GroupBy(x => x.GroupId);
+
+                    foreach (var grouping in groupings)
+                    {
+                        var selectedGroupId = groupIdMappings.First(kv => kv.Value == grouping.Key).Key;
+                        var selectedOptionIds = grouping
+                            .Select(x => optionIdMappings.FirstOrDefault(kv => kv.Value == x.OptionId).Key)
+                            .ToArray();
+
+                        _selectOptions(currentStepId, selectedGroupId, selectedOptionIds);
+                    }
+
+                    // We need to explicitly call this, since our implementations
+                    // return from RequestMultipleChoices when the user wants to go
+                    // to the next step. Once again, the API is ass.
+                    _continueToNextStep(forward: true, currentStepId);
                 }
             });
     }
 
-    private static IEnumerable<ChoiceGroup<int, int>> ToChoices(GroupList groups)
+    private static GuidedInstallationStep ToGuidedInstallationStep(
+        FomodInstaller.Interface.ui.GroupList groups,
+        ICollection<KeyValuePair<int, GroupId>> groupIdMapping,
+        ICollection<KeyValuePair<int, OptionId>> optionIdMappings)
     {
-        return groups.group.Select(group => new ChoiceGroup<int, int>
+        var stepGroups = groups.group.Select(group =>
         {
-            Id = group.id,
-            Type = ConvertChoiceType(group.type),
-            Query = group.name,
-            Options = ToOptions(group.options),
+            var groupId = GroupId.From(Guid.NewGuid());
+            groupIdMapping.Add(new KeyValuePair<int, GroupId>(group.id, groupId));
+
+            return new OptionGroup
+            {
+                Id = groupId,
+                Type = ConvertOptionGroupType(group.type),
+                Description = group.name,
+                Options = ToOptions(group.options, optionIdMappings).ToArray(),
+            };
         });
+
+        return new GuidedInstallationStep
+        {
+            Id = StepId.From(Guid.NewGuid()),
+            Groups = stepGroups.ToArray()
+        };
     }
 
-    private static IEnumerable<Option<int>> ToOptions(IEnumerable<Option> options)
+    private static IEnumerable<Option> ToOptions(
+        IEnumerable<FomodInstaller.Interface.ui.Option> options,
+        ICollection<KeyValuePair<int, OptionId>> optionIdMappings)
     {
-        return options.Select(option => new Option<int>
+        return options.Select(option =>
         {
-            Id = option.id,
-            Name = option.name,
-            Description = option.description,
-            ImageUrl = option.image != null ? AssetUrl.From(option.image) : null,
-            Type = MakeOptionState(option),
-        }).ToArray();
+            var optionId = OptionId.From(Guid.NewGuid());
+            optionIdMappings.Add(new KeyValuePair<int, OptionId>(option.id, optionId));
```

Comment:
The FOMOD library sends us integers. The `IGuidedInstaller` implementation and everything around it shouldn't be specific for that library. As such, I decided to use GUIDs for IDs and use mappings to convert FOMOD library IDs to our internal ID format.

Reply:
If we were to add more installers, then we can always map to and from a GUID. If we were to us IDs, we'd be more restricted.

Reply:
I see, the problem with GUIDs is that running the installer again will generate different IDs.
Right now it's not a problem, but it might be inconvenient if we later might want to keep track of user choices.

I see that the previous code was using generics for the ID type. Perhaps that could be an option.

Otherwise for the purpose of making the id persistent across reinstalls, we could try to generate a unique id from discriminatory data of the option (group+name+ID or something like that), using a hash function. But I guess that would be out of scope for this PR.

Thoughts on using generics?

Reply:
We had generics previously with `Option<TOptionId>` and `OptionGroup<TGroupId, TOptionId>`. This PR removes that. They are a pain to work with, especially when dealing with the VMs. Tracking user choices and making it persistent is out of scope for this PR and #316. If we wanted to add that, we can easily change our code, since the actual value is hidden behind a custom type.